ql/src/java/org/apache/hadoop/hive/ql/exec/tez/TezSessionState.java

```diff
@@ -144,14 +140,17 @@ public String toString() {
     }
   }
 
-  private HiveResources resources;
+  @VisibleForTesting
+  HiveResources resources;
   @JsonProperty("doAsEnabled")
   private boolean doAsEnabled;
   private boolean isLegacyLlapMode;
   private WmContext wmContext;
   private KillQuery killQuery;
 
-  private static final Cache<String, String> shaCache = CacheBuilder.newBuilder().maximumSize(100).build();
+  // a map for jars (with SHA) that have been already localized in the session
+  @VisibleForTesting
+  final Map<String, LocalResource> localizedJarSha = new HashMap<>();
```

Comment:
do we need to maintain a map for localized resources or it's enough to have a Set<> and skip the duplicates?

Reply:
I wanted to use a simply Set<> first but I realized I need to be able to return with the original LocalResouce when it's already added because the callers of these methods already expect that

Reply:
can we return an empty Optional in that case?

Reply:
btw, we already have Set<LocalResource> localizedResources in TezSessionState

Reply:
shouldn't we simply return the same fileName when sha256 matches? in that case, DagUtils.checkPreExisting should take care or dups?

Reply:
the SHA as a map key is an easy way to find a jar with the same content (e.g. so the same jar file present in the same parcel) localized already...how to utilize only a Set<LocalResource> for this purpose? set.contains() looks risky, as we're intentionally trying to prevent localizing the same jar file (which is present in the local filesystem even with different names), so we would end up calling LocalResource.equals / LocalResource.hashCode while looking it up in the set, which most probably considers LocalResource.url, leading to different hashCode, making this optimization fail by adding e.g. hive-jar and hive-exec-3.1.3000.7.2.18.0-334.jar as a LocalResource again

Reply:
see DagUtils.localizeResource 
````
if (checkPreExisting(srcFs, src, dest, conf)) {
   return createLocalResource(destFS, dest, type, LocalResourceVisibility.PRIVATE);
}
````
if the dest is already localized, no need to do it again. <Set> should be completely sufficient

Reply:
okay, let me check